crates/ide/src/references.rs

```diff
@@ -198,6 +202,103 @@ fn get_struct_def_name_for_struct_literal_search(
     None
 }
 
+fn try_find_lifetime_references(
+    syntax: &SyntaxNode,
+    position: FilePosition,
+) -> Option<RangeInfo<ReferenceSearchResult>> {
+    let lifetime_token =
+        syntax.token_at_offset(position.offset).find(|t| t.kind() == SyntaxKind::LIFETIME)?;
+    let parent = lifetime_token.parent();
+
+    match_ast! {
+        match parent {
+            ast::LifetimeArg(it) => find_all_lifetime_references(),
+            // &'lt self
+            ast::SelfParam(it) => find_all_lifetime_references(),
+            ast::RefType(it) => find_all_lifetime_references(),
+            ast::LifetimeParam(it) => find_all_lifetime_references(),// decl
+            // HRTB
+            ast::WherePred(it) => find_all_lifetime_references(),
+            ast::TypeBound(it) => find_all_lifetime_references(),
+            ast::Label(_it) => find_all_label_references(position, lifetime_token, &parent), // decl
+            ast::BreakExpr(_it) => find_all_label_references(position, lifetime_token, &parent),
+            ast::ContinueExpr(_it) => find_all_label_references(position, lifetime_token, &parent),
+            _ => None,
+        }
+    }
```

Comment:
given that we don't make use of well-typed is, we might just `match parent.kind()` and fold some args together.

Reply:
Also, seems like this can be factored a bit better?

```
enum LifetimeClass {
    LifetimeParam, Label
}

impl LifetimeClass {
  fn classify(lt: ast::Lifetime) -> LifeTimeClass. 
}
```